### PR TITLE
Add .npy support to debug_to_file()

### PR DESCRIPTION
**Pull request by steven-johnson**
Built on top of https://github.com/halide/Halide/pull/8175, this adds .npy as an option. This is actually pretty great because it's easy to do something like

```
ss = numpy.load("my_file.npy")
print(ss)
```

in Python and get nicely-formatted output, which can sometimes be a lot easier for debugging that inserting lots of print() statements (see https://github.com/halide/Halide/issues/8176)

Did a drive-by change to the correctness test to use this format instead of .mat.

### CODE REVIEW

File: src/runtime/write_debug_image.cpp [hunk 1]
```diff
@@ -125,6 +128,42 @@ struct ScopedFile {
     }
 };
 
+#if (defined(__BYTE_ORDER) && __BYTE_ORDER == __BIG_ENDIAN) || defined(HALIDE_FORCE_BIG_ENDIAN)
```
Comment (abadams) [inline src/runtime/write_debug_image.cpp:131]: There are parts of the compiler that assume the target is little-endian already, so I don't think you need to worry about big-endian systems in the runtime.

File: src/runtime/write_debug_image.cpp [hunk 2]
```diff
@@ -142,13 +181,14 @@ WEAK extern "C" int halide_debug_to_file(void *user_context, const char *filenam
         return halide_error_code_bad_dimensions;
     }
 
-    if (auto result = halide_copy_to_host(user_context, buf);
-        result != halide_error_code_success) {
+    if (auto result = halide_copy_to_host(user_context, buf); result != halide_error_code_success) {
+        halide_error(user_context, "debug_to_file(): halide_copy_to_host failed\n");
```
Comment (abadams) [inline src/runtime/write_debug_image.cpp:185]: copy_to_host has already called halide_error if it has failed. This happens at the lowest level possible (e.g. in cuda_do_multidimensional_copy) where the most information is available. We shouldn't add a redundant call.

File: src/runtime/write_debug_image.cpp [hunk 3]
```diff
@@ -167,7 +207,75 @@ WEAK extern "C" int halide_debug_to_file(void *user_context, const char *filenam
...
+                di = d.dtype;
+                break;
+            }
+        }
+        if (di.byte_order == 0) {
+            halide_error(user_context, "Unsupported element type for .npy in debug_to_file()\n");
+            return halide_error_code_debug_to_file_failed;
+        }
+
+        constexpr int kMaxDictStringSize = 1024;
```
Comment (abadams) [inline src/runtime/write_debug_image.cpp:223]: We don't typically use this naming convention for constants. max_dict_string_size would be more in keeping with the rest of the compiler.

File: src/runtime/write_debug_image.cpp [hunk 4]
```diff
@@ -167,7 +207,75 @@ WEAK extern "C" int halide_debug_to_file(void *user_context, const char *filenam
...
+        char dict_string_buf[kMaxDictStringSize];
+        char *dst = dict_string_buf;
+        char *end = dict_string_buf + kMaxDictStringSize - 1;
+
+        dst = halide_string_to_string(dst, end, "{'descr': '");
```
Comment (abadams) [inline src/runtime/write_debug_image.cpp:228]: Why not a StringStreamPrinter here?

File: test/correctness/debug_to_file.cpp [hunk 5]
```diff
@@ -15,13 +15,13 @@ int main(int argc, char **argv) {
         return 0;
     }
 
-    std::string f_mat = Internal::get_test_tmp_dir() + "f.mat";
```
Comment (abadams) [inline test/correctness/debug_to_file.cpp:18]: I'm confused - doesn't this remove .mat testing?

**Comment by abadams** [comment]
This function shouldn't call halide_error. Halide_error is called at the Halide IR level if it fails, with a more informative error message (it names the Func)

**Comment by abadams** [comment]
Would be nice to in some way distinguish in the type system which parts of the runtime should return an error code and have already called halide_error, vs which parts return an error code that still needs a call to halide_error to happen above.

File: src/runtime/write_debug_image.cpp (hunk 2 above)
Comment (steven-johnson) [reply src/runtime/write_debug_image.cpp:185]: This didn't get added -- just reformatted by clang-format. I can take it out but it clearly isn't hurting anything...

Comment (steven-johnson) [reply src/runtime/write_debug_image.cpp:185]: Oh wait, nm, I see. Yeah, there were other returns without calling halide_error() that I fixed, but this one is redundant, removing.

File: test/correctness/debug_to_file.cpp (hunk 5 above)
Comment (steven-johnson) [reply test/correctness/debug_to_file.cpp:18]: Yeah, it does so in favor of npy. I could back out this change, or better yet, do both.

**Comment by steven-johnson** [comment]
> Would be nice to in some way distinguish in the type system which parts of the runtime should return an error code and have already called halide_error, vs which parts return an error code that still needs a call to halide_error to happen above.

Yeah, everything about error handling in the runtime is messy. A few years ago I started work on a proposal to rationalize it but ended up abandoning it.

**Comment by steven-johnson** [comment]
> This function shouldn't call halide_error. Halide_error is called at the Halide IR level if it fails, with a more informative error message (it names the Func)

Surely you don't mean that we shouldn't call it anywhere in halide_debug_to_file()?

File: test/correctness/debug_to_file.cpp (hunk 5 above)
Comment (abadams) [reply test/correctness/debug_to_file.cpp:18]: Yeah I think this should test .mat and .npy

**Comment by steven-johnson** [comment]
PTAL

**Comment by abadams** [comment]
I do mean that. It produces an error message with less information than if we just return an error code without calling halide_error, and if halide_error has been overridden to return instead of aborting. It produces two error messages for the same error.

File: src/runtime/write_debug_image.cpp (hunk 4 above)
Comment (steven-johnson) [reply src/runtime/write_debug_image.cpp:229]: ...because I didn't think of it?

**Comment by steven-johnson** [comment]
PTAL -- I can use StackStringPrinter if you prefer but the underlying codegen will likely be identical, LMK

File: src/DebugToFile.cpp [hunk 6]
```diff
@@ -42,6 +42,8 @@ class DebugToFile : public IRMutator {
                 num_elements *= bound.extent;
             }
 
+            // TODO: why do we bother with this? halide_debug_to_file()
```
Comment (abadams) [inline src/DebugToFile.cpp:45]: This code predates halide_buffer_t, and buffer_t only had an elem_size, not a type code. Feel free to fix.

Comment (steven-johnson) [reply src/DebugToFile.cpp:45]: > This code predates halide_buffer_t, and buffer_t only had an elem_size, not a type code. Feel free to fix.

I'll do that as a followup.

**Comment by steven-johnson** [comment]
Gentle Review Ping

**Review by abadams — APPROVED** [review]

**Pull request merged**